analyses/survival-analysis/utils/survival_models.R

```diff
@@ -0,0 +1,618 @@
+      df,
+      ind_var = terms,
+      test    = "cox.reg",
+      metadata_sample_col = "Kids_First_Biospecimen_ID",
+      os_days_col = years_col # we want to use years, not days
+    )
+      } else if (model_type == "multivariate") {
+        
+        if (years_col == "EFS_years"){
+          
+          # Recode OS_Status (for univariate, `survival_analysis()` does the recoding)
+          df <- df %>%
+            mutate(EFS_status = ifelse(EFS_status == "NO EVENT", 0, 1))
+          
+          # Fit model
+          fitted_multi <- survival::coxph(
+            formula(
+              paste0("survival::Surv(time = EFS_years, event = EFS_status) ~ ", terms)
+            ),
+            data = df
+          )
+          
+        } else if (years_col == "OS_years"){
+          
+          # Recode OS_Status (for univariate, `survival_analysis()` does the recoding)
+          df <- df %>%
+            mutate(OS_status = ifelse(OS_status == "LIVING", 0, 1))
+          
+          # Fit model
+          fitted_multi <- survival::coxph(
+            formula(
+              paste0("survival::Surv(time = OS_years, event = OS_status) ~ ", terms)
+            ),
+            data = df
+          )
+          
+        }
+
+    # Set up list object to match parts of `survival_analysis()` output we need
+    fit <- list(
+      model = fitted_multi,
+      table = broom::tidy(fitted_multi)
+    )
+  }
+  
+  # Save model RDS
+  readr::write_rds(fit$model,
+                   # remove spaces in case (common for cancer groups)
+                   str_replace_all(output_file, " ", "-"))
+  
+  # Return the model fit table to be printed out
+  fit$table
+}
+
+
+
+
+
+
+plotKM <- function(model,
+                   variable,
+                   combined = FALSE, 
+                   title) {
+  
+  if (combined == FALSE){
+    
+    if ("OS_days" %in% names(model$original_data)){
+      
+      variable_os <- model$original_data %>% 
+        pull(variable)
+      
+      event_type <- "OS"
+      diff_obj <- survdiff(survival::Surv(OS_days, OS_status) ~ variable_os
+                              , model$original_data)
+      diff_pvalue <- 1 - pchisq(diff_obj$chisq, length(diff_obj$n) - 1)
+      diff_pvalue_formatted <- format(
+        signif(diff_pvalue, 2),
+        scientific = FALSE)
+
+    }
+    
+    if ("EFS_days" %in% names(model$original_data)){
+      
+      event_type <- "EFS"
+      
+      variable_efs <- model$original_data %>% 
+        pull(variable)
+      
+      diff_obj <- survdiff(survival::Surv(EFS_days, EFS_status) ~ variable_efs,  
+                           model$original_data)
+      diff_pvalue <- 1 - pchisq(diff_obj$chisq, length(diff_obj$n) - 1)
+      diff_pvalue_formatted <- format(
+        signif(diff_pvalue, 2),
+        scientific = FALSE)
+      
+    }
+    
+    
+    levels <- model$original_data %>%
+      pull(variable) %>%
+      unique()
+    levels <- levels[order(levels)]
```

Comment:
```suggestion
    levels <- levels[order(levels)][!is.na(levels)]
```

This would resolve the current error, if we wanted to move forward to plotting cases where variable value is NA. This is actually a step that's already performed when `combined == TRUE`, so I must have missed it here.

Reply:
Thank you so much! Change has been made.